File: scripts/maintenance/update_sha_hashes.py
#!/usr/bin/env python3
"""
SHA-256 Hash Update Script

This script automatically updates SHA-256 hashes in workflow files when
the corresponding source files change. It ensures that security verification
hashes stay in sync with file modifications.

Usage:
    python3 update_sha_hashes.py [--dry-run] [--verbose]

METADATA:
    AUTHOR: Moko Consulting LLC
    COPYRIGHT: 2026 Moko Consulting LLC
    LICENSE: GPL-3.0-or-later
    VERSION: 03.01.01
    CREATED: 2026-02-02
    UPDATED: 2026-02-02
"""

import argparse
import hashlib
import re
import sys
from pathlib import Path
from typing import Dict, List, Tuple


# Configuration: Map source files to their workflow locations
SHA_MAPPINGS = [
    {
        'source_file': 'scripts/validate/validate_codeql_config.py',
        'workflow_file': '.github/workflows/standards-compliance.yml',
        'pattern': r'EXPECTED_SHA256="([a-f0-9]{64})"',
        'description': 'CodeQL configuration validator'
    }
]


def calculate_sha256(file_path: Path) -> str:
    """Calculate SHA-256 hash of a file.
    
    Args:
        file_path: Path to the file
        
    Returns:
        Hexadecimal SHA-256 hash string
    """
    sha256_hash = hashlib.sha256()
    with open(file_path, 'rb') as f:
        # Read in chunks to handle large files efficiently
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()


def update_sha_in_file(
    workflow_file: Path,
    pattern: str,
    new_hash: str,
    dry_run: bool = False
) -> Tuple[bool, str, str]:
    """Update SHA-256 hash in workflow file.
    
    Args:
        workflow_file: Path to workflow file
        pattern: Regex pattern to find the hash
        new_hash: New hash value to set
        dry_run: If True, don't actually write changes
        
    Returns:
        Tuple of (changed, old_hash, new_hash)
    """
    with open(workflow_file, 'r') as f:
        content = f.read()
    
    # Find current hash
    match = re.search(pattern, content)
    if not match:
        return False, '', new_hash
    
    old_hash = match.group(1)
    
    # Check if update needed
    if old_hash == new_hash:
        return False, old_hash, new_hash
    
    # Replace hash
    new_content = re.sub(pattern, f'EXPECTED_SHA256="{new_hash}"', content)
    
    if not dry_run:
        with open(workflow_file, 'w') as f:
            f.write(new_content)
    
    return True, old_hash, new_hash


def main():
    """Main execution function."""
    parser = argparse.ArgumentParser(
        description='Update SHA-256 hashes in workflow files'
    )
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Show what would be changed without making changes'
    )
    parser.add_argument(
        '--verbose',
        action='store_true',
        help='Show detailed output'
    )
    args = parser.parse_args()
    
    # Get repository root
    repo_root = Path(__file__).parent.parent.parent
    
    print("🔐 SHA-256 Hash Update Script")
    print("=" * 60)
    
    if args.dry_run:
        print("🔍 DRY RUN MODE - No changes will be made")
        print("=" * 60)
    
    changes_made = False
    errors = []
    
    for mapping in SHA_MAPPINGS:
        source_path = repo_root / mapping['source_file']
        workflow_path = repo_root / mapping['workflow_file']
        
        print(f"\n📄 Processing: {mapping['description']}")
        print(f"   Source: {mapping['source_file']}")
        
        # Check if source file exists
        if not source_path.exists():
            error_msg = f"❌ Source file not found: {source_path}"
            print(error_msg)
            errors.append(error_msg)
            continue
        
        # Check if workflow file exists
        if not workflow_path.exists():
            error_msg = f"❌ Workflow file not found: {workflow_path}"
            print(error_msg)
            errors.append(error_msg)
            continue
        
        # Calculate current hash
        current_hash = calculate_sha256(source_path)
        if args.verbose:
            print(f"   Current SHA-256: {current_hash}")
        
        # Update workflow file
        changed, old_hash, new_hash = update_sha_in_file(
            workflow_path,
            mapping['pattern'],
            current_hash,
            args.dry_run
        )
        
        if changed:
            changes_made = True
            print(f"   ✅ Hash updated in: {mapping['workflow_file']}")
            print(f"      Old: {old_hash}")
            print(f"      New: {new_hash}")
            if args.dry_run:
                print(f"      (Would be updated - dry run mode)")
        else:
            if old_hash == new_hash:
                print(f"   ℹ️  Hash already up to date")
            else:
                print(f"   ⚠️  No hash found to update")
    
    print("\n" + "=" * 60)
    
    if errors:
        print(f"❌ Completed with {len(errors)} error(s)")
        for error in errors:
            print(f"   {error}")
        return 1
    
    if changes_made:
        if args.dry_run:
            print("✅ Changes detected (dry run - no files modified)")
        else:
            print("✅ All SHA-256 hashes updated successfully")
        return 0
    else:
        print("ℹ️  No changes needed - all hashes up to date")
        return 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/maintenance/update_sha_hashes.php
#!/usr/bin/env php
<?php
/**
 * Copyright (C) 2026 Moko Consulting <hello@mokoconsulting.tech>
 *
 * This file is part of a Moko Consulting project.
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * FILE INFORMATION
 * DEFGROUP: MokoStandards.Scripts.Maintenance
 * INGROUP: MokoStandards
 * REPO: https://github.com/mokoconsulting-tech/MokoStandards
 * PATH: /scripts/maintenance/update_sha_hashes.php
 * VERSION: 04.00.00
 * BRIEF: Update SHA-256 hashes in script registry
 */

declare(strict_types=1);

/**
 * Script Registry Hash Updater
 * 
 * Updates SHA-256 hashes for all scripts in the registry
 */
class ScriptRegistryUpdater
{
    private const REGISTRY_PATH = 'scripts/.script-registry.json';
    
    private bool $dryRun = false;
    private bool $verbose = false;
    private array $changes = [];
    
    public function __construct(array $args)
    {
        $this->parseArguments($args);
    }
    
    private function parseArguments(array $args): void
    {
        foreach ($args as $arg) {
            if ($arg === '--dry-run') {
                $this->dryRun = true;
            } elseif ($arg === '--verbose' || $arg === '-v') {
                $this->verbose = true;
            } elseif ($arg === '--help' || $arg === '-h') {
                $this->showHelp();
                exit(0);
            }
        }
    }
    
    private function showHelp(): void
    {
        echo "Usage: php update_sha_hashes.php [OPTIONS]\n\n";
        echo "Options:\n";
        echo "  --dry-run     Check for changes without updating the registry\n";
        echo "  --verbose     Show detailed output\n";
        echo "  --help        Show this help message\n";
        echo "\n";
    }
    
    public function run(): int
    {
        try {
            $this->log("🔐 SHA-256 Hash Update Tool", true);
            $this->log(str_repeat("=", 50), true);
            
            if ($this->dryRun) {
                $this->log("Mode: DRY RUN (no changes will be made)", true);
            }
            
            // Load registry
            $registry = $this->loadRegistry();
            
            // Update hashes
            $updatedRegistry = $this->updateHashes($registry);
            
            // Save if not dry run and there are changes
            if (!$this->dryRun && !empty($this->changes)) {
                $this->saveRegistry($updatedRegistry);
                $this->log("\n✅ Registry updated successfully", true);
            } elseif (empty($this->changes)) {
                $this->log("\nℹ️  No changes needed - all hashes are current", true);
            } else {
                $this->log("\n✅ Dry run complete - changes detected but not applied", true);
            }
            
            return 0;
            
        } catch (Exception $e) {
            fwrite(STDERR, "❌ Error: " . $e->getMessage() . "\n");
            return 1;
        }
    }
    
    private function loadRegistry(): array
    {
        if (!file_exists(self::REGISTRY_PATH)) {
            throw new Exception("Registry file not found: " . self::REGISTRY_PATH);
        }
        
        $content = file_get_contents(self::REGISTRY_PATH);
        $registry = json_decode($content, true);
        
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new Exception("Failed to parse registry JSON: " . json_last_error_msg());
        }
        
        $this->log("Registry loaded: " . count($registry['scripts']) . " scripts tracked");
        
        return $registry;
    }
    
    private function updateHashes(array $registry): array
    {
        $this->log("\nChecking scripts for changes...\n");
        
        foreach ($registry['scripts'] as $index => &$script) {
            $path = $script['path'];
            
            if (!file_exists($path)) {
                $this->log("⚠️  Skipping missing file: {$path}");
                continue;
            }
            
            // Calculate current hash
            $currentHash = hash_file('sha256', $path);
            $currentSize = filesize($path);
            
            // Check if changed
            if ($currentHash !== $script['sha256']) {
                $this->changes[] = [
                    'path' => $path,
                    'old_hash' => $script['sha256'],
                    'new_hash' => $currentHash,
                ];
                
                $this->log("🔄 Hash updated: {$path}", true);
                
                if ($this->verbose) {
                    $this->log("   Old: {$script['sha256']}");
                    $this->log("   New: {$currentHash}");
                }
                
                // Update in registry
                $script['sha256'] = $currentHash;
                $script['size_bytes'] = $currentSize;
            } else {
                $this->log("✓ No change: {$path}");
            }
        }
        
        // Update metadata timestamp if there are changes
        if (!empty($this->changes)) {
            $microtime = microtime(true);
            $dt = DateTime::createFromFormat('U.u', sprintf('%.6f', $microtime), new DateTimeZone('UTC'));
            $registry['metadata']['generated_at'] = $dt->format('Y-m-d\TH:i:s.u\Z');
        }
        
        $this->log("\nSummary:");
        $this->log("  Total scripts: " . count($registry['scripts']), true);
        $this->log("  Changed: " . count($this->changes), true);
        
        return $registry;
    }
    
    private function saveRegistry(array $registry): void
    {
        $json = json_encode($registry, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
        
        if ($json === false) {
            throw new Exception("Failed to encode registry JSON: " . json_last_error_msg());
        }
        
        if (file_put_contents(self::REGISTRY_PATH, $json) === false) {
            throw new Exception("Failed to write registry file");
        }
        
        $this->log("Registry saved: " . self::REGISTRY_PATH);
    }
    
    private function log(string $message, bool $force = false): void
    {
        if ($this->verbose || $force) {
            echo $message . "\n";
        }
    }
}

// Run the updater
$updater = new ScriptRegistryUpdater(array_slice($argv, 1));
exit($updater->run());
